design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
   name: example
 spec:
   compositeTypeRef:
-    apiVersion: database.example.org/v1alpha1
+    apiVersion: database.example.org/v1
     kind: XPostgreSQLInstance
   functions:
-  - name: my-cool-function
-    type: Container
-    # Configuration specific to `type: Container` functions.
-    container:
-      # The OCI image to pull and run.
-      image: xkpg.io/my-cool-function:0.1.0
-      # Whether to pull the function Never, Always, or IfNotPresent.
-      imagePullPolicy: IfNotPresent
-      # Secrets used to pull from a private registry.
-      imagePullSecrets:
-      - namespace: crossplane-system
-        name: my-xpkg-io-creds
-      # Note that only resource limits are supported - not requests.
-      # The function will be run with the specified resource limits.
-      resources:
-        limits:
-          memory: 64Mi
-          cpu: 250m
-      # Defaults to 'Isolated' - i.e an isolated network namespace.
-      network: Accessible
-      # How long the function may run before it's killed. Defaults to 10s.
-      timeout: 30s
-      # Containers are run by an external process listening at the supplied
-      # endpoint. Specifying an endpoint is optional; the endpoint defaults to
-      # the below value.
-      runner:
-        endpoint: unix:///@crossplane/fn/default.sock
-    # An x-kubernetes-embedded-resource RawExtension (i.e. an unschemafied
-    # Kubernetes resource). Passed to the function as the config block of its
-    # FunctionIO.
+  - name: compose-xr-using-go-templates
+    functionRef:
+      name: go-templates
     config:
-      apiVersion: database.example.org/v1alpha1
-      kind: Config
-      metadata:
-        name: cloudsql
-      spec:
-        version: POSTGRES_9_6
+      apiVersion: example.org/v1
+      kind: GoTemplate
+      source: Inline  # Or Remote, if you want to pull templates from git.
+      inline: |
+        {{- range $i := until ( . desired.composite.resource.spec.count ) }}
+        ---
+        apiVersion: iam.dummy.upbound.io/v1alpha1
+        kind: Robot
+        metadata:
+          annotations:
+            crossplane.io/external-name: {{ .desired.composite.resource | getExternalName }}-robot-{{ $i }}
+        spec:
+          forProvider:
+          {{- if $i | printf "robot-%d" | getComposedResource }}
```

Comment:
how does the template rendering runtime know where this `getComposedResource` function lives and how to call it? 🤔

Reply:
Keep in mind this is just a sketch, not something I've _fully_ worked out. 😄 

That said, I think the way this works is that the author of the Go Templating Composition Function would provide a set of template functions that could be used in templates.

Helm has a built-in set for example - https://helm.sh/docs/chart_template_guide/function_list. Many of them come from https://github.com/Masterminds/sprig, which is kind of a little stdlib for go-templates. Others are specific to Helm. I imagine we'd do the same - include sprig, and also write a few things like this that made life easier for folks writing Compositions.

I don't think these are easily discoverable or self-documenting, so we'd have to do like Helm does and write documentation that explains how this Function works. I imagine that would likely end up at https://marketplace.upbound.io and/or in the Function's README.